### PR TITLE
Implement readString

### LINKED ISSUE #606: Implement `BufReader.readString(delim)`
Noticed that this is [not yet implemented](https://github.com/denoland/deno_std/blob/a8f6cf7b4fa4030878bb702a65f28f469ee1f87d/io/bufio.ts#L215). Should be a very useful utility (also since both `readSlice` and `readLine` are implemented but in the documentation pointing to prefer using `readString`).

Also may be a good task for new contributors

### PR DESCRIPTION
**Pull request by sh7dm**
Closes #606

### CODE REVIEW

File: io/bufio.ts [hunk 1]
```diff
@@ -212,7 +212,8 @@ export class BufReader implements Reader {
    * For simple uses, a Scanner may be more convenient.
    */
   async readString(_delim: string): Promise<string | Deno.EOF> {
```
Comment (kitsonk) [inline io/bufio.ts]: The leading `_` should be removed from this now... it was a signal to allow unused arguments in the signature.

Comment (sh7dm) [reply io/bufio.ts]: @kitsonk done, thanks.

File: io/bufio.ts [hunk 2]
```diff
@@ -211,8 +211,9 @@ export class BufReader implements Reader {
    * delim.
    * For simple uses, a Scanner may be more convenient.
    */
-  async readString(_delim: string): Promise<string | Deno.EOF> {
-    throw new Error("Not implemented");
+  async readString(delim: string): Promise<string | Deno.EOF> {
+    const buffer = await this.readSlice(delim.charCodeAt(0));
```
Comment (ry) [inline io/bufio.ts]: Maybe throw if delim.length != 1

**Comment by kevinkassimo** [comment]
Hmm I am also thinking if we could provide a version that possibly allow multi-char delimiter?

**Comment by zekth** [comment]
> Hmm I am also thinking if we should also provide a version that possibly allow multi-char delimiter?

IMO we need multichar delimiter

**Comment by kevinkassimo** [comment]
> > Hmm I am also thinking if we should also provide a version that possibly allow multi-char delimiter?
> 
> IMO we need multichar delimiter

On second thought I noticed that since BufReader is implemented in a way strictly compliant with Go’s bufio counterparts, maybe single char restriction is fine here. I do hope though if we can provide another utility for unlimited buffered reader that automatically resize/reallocate the internal buffer which would allow implementation of multi-char delim easier (maybe an extension of `deno.Buffer`. We have something similar for xeval that recently got moved here)

File: io/bufio.ts (hunk 2 above)
Comment (sh7dm) [reply io/bufio.ts]: Keeping in mind comments :point_down:, I'll check if `delim.lenght > 0`

File: io/bufio.ts [hunk 3]
```diff
@@ -211,8 +211,11 @@ export class BufReader implements Reader {
    * delim.
    * For simple uses, a Scanner may be more convenient.
    */
-  async readString(_delim: string): Promise<string | Deno.EOF> {
-    throw new Error("Not implemented");
+  async readString(delim: string): Promise<string | Deno.EOF> {
+    if (delim.length !== 1)
+      throw new Error("Delimiter should be a single character");
+    const buffer = await this.readSlice(delim.charCodeAt(0));
+    return new TextDecoder().decode(buffer || undefined);
```
Comment (sh7dm) [inline io/bufio.ts]: should we use `strings` module here?

Comment (ry) [reply io/bufio.ts]: No this is fine

**Pull request merged**